### DISCUSSION TITLE
JTT 808 module support

### DISCUSSION THREAD

**Discussion by vhlukhotskyy** [emqx/emqx#4716] [Q&A]
I am using EMQX EE 4.2.6 running as a single node on Ubuntu 20.04 server with the default configuration and JTT 808 module installed using this [guide](https://docs.emqx.io/en/enterprise/v4.2/modules/jt808_protocol.html#create-module) 

What I currently trying to achieve is to send basically any JTT 808 message to the EMQX and receive by subscribing to jtt808/# 
topic.

- JTT 808 client - https://github.com/yezhihao/jt808-server (test application in tests)
- MQTT client - Eclipse Mosquitto  2.0.10

So far I cannot see any messages received by the mqtt client.

Would be great if you could help me. 
Thanks

**Comment by k32** [comment]
Hi,

Can you reproduce the issue with debug logs enabled (as specified here: https://docs.emqx.io/en/enterprise/v4.2/configuration/configuration.html#log-level) and share the logs?

**Reply by vhlukhotskyy** [reply]
Hi @k32 
Thanks for a quick reply. 

Please find the log file in the attachments
[emqx.log.2.txt](https://github.com/emqx/emqx/files/6399358/emqx.log.2.txt)

**Reply by k32** [reply]
Thanks for the logs. Disclaimer: I have absolutely zero knowledge of JT808 protocol, but purely by looking at the code, there should be "phone" field in the client info, that is missing.

**Reply by k32** [reply]
P.S. Perhaps you need to restart EMQ X to enable debug logs.

**Reply by vhlukhotskyy** [reply]
Same here. Could be a problem with the JT808 client in the end. 
I am not sure exactly about the "phone" field though since from what I see in the logs is:
2021-04-29 15:26:21.646 [debug] [JT808 Proto] RECV #{<<"body">> =>
                         #{<<"city">> => 1,<<"color">> => 0,
                           <<"dev_id">> => <<"0648632">>,
                           <<"license_number">> => <<"123456">>,
                           <<"manufacturer">> => <<"12345">>,
                           <<"model">> =>
                               <<84,87,49,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0>>,
                           <<"province">> => 1},
                     <<"header">> =>
                         #{<<"encrypt">> => 0,<<"len">> => 43,
                           <<"msg_id">> => 256,<<"msg_sn">> => 754,
                           <<"phone">> => <<"053810648632">>}}
It seems like "phone" is passed in the message header.

**Reply by vhlukhotskyy** [reply]
Another strange thing is that the message content sent by the JT808 client seems to be totally different from what is displayed in the broker logs. 
[messageId=256, properties=0, isVersion=false, versionNo=0, isSubpackage=false, mobileNo=12345678901, serialNo=32767],T0100[provinceId=31,cityId=115,makerId=4,deviceModel=BSJ-GF-06,deviceId=test123,plateColor=1,plateNo=测A888888]

**Reply by k32** [reply]
Sorry, I was searching for "phone" in the truncated version of your logs

**Reply by vhlukhotskyy** [reply]
Here are the newer logs:
[emqx.log.3.txt](https://github.com/emqx/emqx/files/6400052/emqx.log.3.txt)

**Reply by vhlukhotskyy** [reply]
> Another strange thing is that the message content sent by the JT808 client seems to be totally different from what is displayed in the broker logs.
> [messageId=256, properties=0, isVersion=false, versionNo=0, isSubpackage=false, mobileNo=12345678901, serialNo=32767],T0100[provinceId=31,cityId=115,makerId=4,deviceModel=BSJ-GF-06,deviceId=test123,plateColor=1,plateNo=测A888888]

Nevermind this. Turned out to be another device sending messages.

**Reply by k32** [reply]
I assume you're running a test service? Is it possible to disable the other device sending data? Perhaps are two problems at the same time. I see two kinds of errors: 1) crashes due to lack of phone number 2) authentication failures due to "no_schema". It would be easier to deal with one at the time.

**Reply by k32** [reply]
P.S. Authentication fails because I see that the registray parameter is an empty string: `{registray,<<>>},` It should be a URL

**Reply by vhlukhotskyy** [reply]
Yes, sure. The first one is the test application and the second one the real device sending messages. 
Let's go with the authentication failures, since the device connectivity was the problem in the first place.

**Reply by k32** [reply]
Can you share contents of `etc/plugins/emqx_jt808.conf` file?

**Reply by vhlukhotskyy** [reply]
> P.S. Authentication fails because I see that the registray parameter is an empty string: `{registray,<<>>},` It should be a URL

Yes, but the configuration hint says that Registray URL as well as Authorization URL are mandatory only when we have  Allow Anonymous = true, which is not the case. 

By the way, is the Registray URL and Authorization URL some kind of web hook and the endpoints itself should be provided as part of the business logic application?

**Reply by vhlukhotskyy** [reply]
> Can you share contents of `etc/plugins/emqx_jt808.conf` file?

[emqx_jt808.conf.txt](https://github.com/emqx/emqx/files/6400339/emqx_jt808.conf.txt)

**Reply by k32** [reply]
Thanks. Interesting, the registray parameter is not enabled, but EMQX read it as an empty string instead of atom `undefined`. This could be the root cause.

**Reply by k32** [reply]
Can you run `application:get_all_env(emqx_jt808).` in the EMQ X console and show the output?

**Reply by vhlukhotskyy** [reply]
> Can you run `application:get_all_env(emqx_jt808).` in the EMQ X console and show the output?

Sorry for the long reply. 
The output is just an empty array:
(emqx@127.0.0.1)2> application:get_all_env(emqx_jt808).
[]

**Reply by k32** [reply]
Thanks for the info. For some reason I can't reproduce the problem with the empty environment for `emqx_jt808` app using `emqx-ee` docker container from docker hub. Could you please try unloading and loading the jt808 plugin and repeating `get_all_env` command?

**Reply by vhlukhotskyy** [reply]
I have removed the JT 808 module and have added it once again. 
Same result for the get_all_env command. 

FYI
I have installed EMQX EE from a .deb package

**Reply by k32** [reply]
Thanks for the info. Unfortunately, I can't reproduce this behavior with `.deb` file install either. Could you share more details about your setup? Which configuration changes did you apply, how do you run emqx, etc.

Also, could you run
```
application:set_env(emqx_jt808, proto, [{up_topic,"jt808/%c/up"},
                                        {dn_topic,"jt808/%c/dn"},
                                        {allow_anonymous,true}]).
``` 
command in the Erlang shell, restart the plugin and try checking if `[error] [JT808 Proto] Register failed: {no_scheme}` still happens?

**Reply by vhlukhotskyy** [reply]
I am using `AWS EC2 t2.micro` instance with `Ubuntu 20.04 LTS` to run the `EMQX EE 4.2.6` installed following this [instructions](https://www.emqx.io/downloads) as `.deb` file. 
I am running the broker as a system service using `sudo emqx start`
Security group for the EC2 instance defines all TCP and UDP port accessible to the public. 

The only change done to the default configuration is JT 808 module install as per [instruction](https://docs.emqx.io/en/enterprise/v4.2/modules/jt808_protocol.html#create-module)

The real device is used to connect to the broker using JT 808 protocol, which was previously tested on manufacturer provided application, that works fine. 

I'll get back to you on the `application:set_env `command result in short

**Reply by k32** [reply]
Thanks for the instruction. My mistake was that I used plugin instead of the module. I think I managed to reproduce the issue.

**Reply by vhlukhotskyy** [reply]
Regarding the `application:set_env` command:
The env for emqx_jt808 is updated, but the issue persists

**Reply by vhlukhotskyy** [reply]
BTW, 
I have tried out JT 808 plugin instead of the module and it works. 
Is there any way I can find examples of JT 808 messages in JSON format?

**Reply by k32** [reply]
Hello,

I will merge the fix soon.

>  I have tried out JT 808 plugin instead of the module and it works.

We're planning to move away from the modules in the future releases, and add REST api for plugin configuration to unify the concepts.

>  Is there any way I can find examples of JT 808 messages in JSON format?

Unfortunately, I am not exactly familiar with JT808 protocol, but maybe someone else will answer.

**Reply by vhlukhotskyy** [reply]
> We're planning to move away from the modules in the future releases, and add REST api for plugin configuration to unify the concepts.
 
 Got it. I thought it was the other way around.
 Will go with the plugin then
 
 Thanks
